lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+			}
+		}
+		validatorsByStake, err = utxoView.GetAllSnapshotValidatorSetEntriesByStakeAtEpochNumber(epochEntrySnapshotAtEpochNumber)
+		if err != nil {
+			return errors.Wrapf(err, "processOrphanBlockPoS: Problem getting validator set at snapshot at epoch number %d", epochEntrySnapshotAtEpochNumber)
+		}
+	} else {
+		// This block is in the current epoch!
+		// First we validate the proposer vote partial signature
+		snapshotAtEpochNumber, err := utxoView.GetCurrentSnapshotEpochNumber()
+		if err != nil {
+			return errors.Wrap(err, "processOrphanBlockPoS: Problem getting current snapshot epoch number")
+		}
+		// Validate the proposer partial sig.
+		isValidPartialSig, err := utxoView.hasValidProposerPartialSignaturePoS(block, snapshotAtEpochNumber)
+		if err != nil {
+			return errors.Wrap(err, "processOrphanBlockPoS: Problem validating proposer partial sig")
+		}
+		if !isValidPartialSig {
+			// We can mark it as ValidateFailed. We'll never accept this block.
+			if _, innerErr := bc.storeValidateFailedBlockInBlockIndex(block); innerErr != nil {
+				return errors.Wrap(innerErr, "processOrphanBlockPoS: Problem adding validate failed block to block index")
+			}
+			return nil
+		}
+		// Next we validate that the leader is correct. We can only do this if the block
+		// is in the current epoch since we need the current epoch entry's initial view
+		// to compute the proper leader.
+		var isBlockProposerValid bool
+		isBlockProposerValid, err = utxoView.hasValidBlockProposerPoS(block)
+		if err != nil {
+			return errors.Wrapf(err, "processOrphanBlockPoS: Problem validating block proposer")
+		}
+		if !isBlockProposerValid {
+			if _, innerErr := bc.storeValidateFailedBlockInBlockIndex(block); innerErr != nil {
+				return errors.Wrapf(innerErr, "processOrphanBlockPoS: Problem adding validate failed block to block index: %v", err)
+			}
+			return nil
+		}
+		// If we get here, we know we have the correct block proposer. We now fetch the validators ordered by
+		// stake so we can validate the QC.
+		validatorsByStake, err = utxoView.GetAllSnapshotValidatorSetEntriesByStake()
+		if err != nil {
+			return errors.Wrap(err, "processOrphanBlockPoS: Problem getting validator set")
+		}
+	}
+	// Okay now we have the validator set ordered by stake, we can validate the QC.
+	if err = bc.isValidPoSQuorumCertificate(block, validatorsByStake); err != nil {
+		// If we hit an error, we know that the QC is invalid and we'll never accept this block,
+		// so we mark it as validate failed.
+		if _, innerErr := bc.storeValidateFailedBlockInBlockIndex(block); innerErr != nil {
+			return errors.Wrapf(innerErr, "processOrphanBlockPoS: Problem adding validate failed block to block index: %v", err)
+		}
+		return nil
+	}
+	// Add to blockIndexByHash with status STORED only as we are not sure if it's valid yet.
+	_, err = bc.storeBlockInBlockIndex(block)
```

Comment:
OK, so as I read through this function I saw a lot of storeValidateFailedBlockInBlockIndex, which again I think is dangerous because it means someone can spam us really hard with invalid blocks. But I think it's a relatively easy fix. Below is a proposal:
* I think we shouldn't store anything until we've done the QC check. On the bright side, it looks like we can do the QC without having to do anything else.
* Given the above, I recommend we break all of the code above storeBlockInBlockIndex into its own function called something like `validateBlockQC` or something like that. When you do this, replace all instances of storeValidateFailedBlockInBlockIndex with returning an error.
* Once you've done that, I would call validateBlockQC at the tippy top of processBlockPoS, effectively as an anti-spam measure. I would do it before you even check the lineage.
* After you've called validateBlockQC, you're now safe to store everything thing you want thereafter because passing the QC check is hard. So you can store the invalid block if it fails the lineage check in processBlockPoS, for example.